design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+ 
+Providing `FORMAT` will set both the key and value formats. Providing `FORMAT` along with either 
+`KEY_FORMAT` or `VALUE_FORMAT` will result in an error.
+
+The key format will follow the same inheritance rules as the current value format. Namely: any 
+derived stream will inherit the format of its leftmost source, unless the format is explicitly set
+in the with clause.
+
+For example:
+
+```sql
+-- Creates a table over a changelog topic with AVRO key and JSON value:
+CREATE TABLE USERS (
+    ID BIGINT PRIMARY KEY,
+    NAME STRING
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    KEY_FORMAT='AVRO',
+    VALUE_FORMAT='JSON'
+  );
+
+-- Creates a stream over a topic with JSON key and value:
+CREATE STREAM BIDS (
+    ITEM_ID BIGINT KEY,
+    USER_ID BIGINT,
+    AMMOUNT INT
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    FORMAT='JSON'
+  );
+
+-- Change the key format of a stream:
+CREATE STREAM AVRP_BIDS 
+  WITH WITH (
+    KEY_FORMAT='AVRO'
+  ) AS
+    SELECT * FROM BIDS;
+
+-- Creates an enriched stream. The key format is inherited from the leftmost source, i.e. JSON:
+CREATE STREAM ENRICHED_BIDS AS 
+  SELECT *
+  FROM BIDS 
+   JOIN USERS ON BIDS.USER_ID = USERS.ID;
+```
+
+For formats that support integration with the schema registry, the key schema will be read and 
+registered with the Schema Registry as needed, following the same pattern as the value schema in 
+the current product.
+
+### New server config
+
+In addition to new `CREATE` properties, the user will also be able to set default key and value 
+formats using system configuration. 
+
+`KEY_FORMAT` will not be a required property _if_ `ksql.persistence.default.format.key` is set. 
+`VALUE_FORMAT` is currently a required property, but will no longer be required _if_ 
+`ksql.persistence.default.format.value` is set.
+
+Somewhat unrelated, but small, we propose also changing the `KAFKA_TOPIC` property from _required_ to 
+_optional_ in `CREATE STREAM` and `CREATE TABLE` statements. If not supplied, it will default to the 
+uppercase name of the stream or table.  This matches the behaviour of the property in 
+`CREATE STREAM AS SELECT` and `CREATE TABLE AS SELECT` statements, where it is already optional.
+
+To additional configurations will be added to control topic names:
+
+* `ksql.persistence.topic.name.case`: if `lower`, defaults ksql to using lowercase topics names
+  where topic names are not explicitly provided. This includes matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic. Default: `upper`.
+* `ksql.persistence.topic.name.case.insensitive`: if `true`, matching a `CREATE TABLE` or 
+  `CREATE STREAM` statement to an existing topic will be case insensitive. Multiple matches will 
+  result in an error. Default: `false`.
+
+With these changes there will no longer be any required `CREATE` properties. This means the `WITH` 
+clause for `CREATE` statements for streams and tables would be optional. We propose this makes the 
```

Comment:
I suppose we still need `partitions` to indicate that the topic should be created, unless we want to re-intorudce a default there

Reply:
I had thought the same.  Yes, we could include a default partition count.  I don't think adding one would cause any problems with the current 'use existing' vs 'create new' logic.

Reply:
I'll add this to the KLIP I'm raising for making the `WITH` clause optional

Reply:
KLIP for making `WITH` clause optional: https://github.com/confluentinc/ksql/pull/6065